### PR TITLE
Implement `ixdtf` for `CustomTimeZone` and `CustomZonedDateTime`

### PR DESCRIPTION
**Pull request by nekevss**
This PR is related to #5201.

It adds the below functionality to `icu_timezone`:

- ~~TryFrom<UtcOffsetRecord> for `GmtOffset`~~
- IXDTF methods for `CustomTimeZone`
- IXDTF methods for `CustomZonedDateTime`
- Adds `new_with_bcp47_id` method to `CustomTimeZone`

Currently, the parsing doesn't resolve any inconsistency that may exist between a DateTime UTC Offset and a IANA identifier time zone annotation, going off the documentation comment that `CustomTimeZone` does not enforce consistency across all fields.

### CODE REVIEW

File: components/timezone/src/ixdtf.rs [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    type Error = InvalidOffsetError;
+
+    fn try_from(value: UTCOffsetRecord) -> Result<Self, Self::Error> {
+        let hour_seconds = i32::from(value.hour) * 3600;
+        let minute_seconds = i32::from(value.minute) * 60;
+        Self::try_from_offset_seconds(
+            i32::from(value.sign as i8) * (hour_seconds + minute_seconds + i32::from(value.second)),
+        )
+    }
+}
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:69]: Nit, here and elsewhere: In https://github.com/unicode-org/icu4x/pull/5260#discussion_r1682820501 we decided not to publicly export these direct conversion functions, at least not until ixdtf 1.0. I think just keep them as private functions. But, if you think these are useful on their own, feel free to make a case for keeping them.

File: components/timezone/src/ixdtf.rs [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+// ==== CustomTimeZone methods and traits ====
+
+impl CustomTimeZone {
+    /// Create a [`CustomTimeZone`] from an IXDTF syntax string.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:76]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo features.*
```

File: components/timezone/src/ixdtf.rs [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+    /// ```
+    ///
+    #[cfg(feature = "compiled_data")]
+    pub fn try_from_ixdtf_str(ixdtf_str: &str) -> Result<Self, ParseError> {
+        Self::try_from_ixdtf_utf8(ixdtf_str.as_bytes())
+    }
+
+    /// Create a [`CustomTimeZone`] from IXDTF syntax utf8 bytes.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:196]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo features.*
```

File: components/timezone/src/ixdtf.rs [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+}
+
+// ==== CustomTimeZone methods and traits ====
+
+impl CustomTimeZone {
+    /// Create a [`CustomTimeZone`] from an IXDTF syntax string.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
+    ///
+    /// # Examples
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:78]: Praise: Good docs tests covering a lot of corner cases

File: components/timezone/src/ixdtf.rs [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+        Ok(Self {
+            date,
+            time,
+            zone: time_zone,
+        })
+    }
+}
+
+impl CustomZonedDateTime<AnyCalendar> {
+    /// Create a [`CustomZonedDateTime`] in ISO-8601 calendar from an IXDTF syntax string.
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:344]:
```suggestion
    /// Create a [`CustomZonedDateTime`] in any calendar from an IXDTF syntax string.
```

File: components/timezone/src/ixdtf.rs [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:346]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo features.*
```

File: components/timezone/src/ixdtf.rs [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        }
+    }
+}
+
+// ==== CustomZonedDateTime methods and traits ====
+
+impl CustomZonedDateTime<Iso> {
+    /// Create a [`CustomZonedDateTime`] in ISO-8601 calendar from an IXDTF syntax string.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:277]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo features.*
```

File: components/timezone/src/ixdtf.rs [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    ///
+    /// For more information on time zone parsing, see [`CustomTimeZone::try_from_ixdtf_str`].
+    #[cfg(feature = "compiled_data")]
+    pub fn try_iso_from_ixdtf_str(ixdtf_str: &str) -> Result<Self, ParseError> {
+        Self::try_iso_from_ixdtf_utf8(ixdtf_str.as_bytes())
+    }
+
+    /// Create a [`CustomZonedDateTime`] in ISO-8601 calendar from IXDTF syntax utf8 bytes.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:313]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo feature.*
```

File: components/timezone/src/ixdtf.rs [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+    ///     gmt_offset: Some(GmtOffset::try_from_offset_seconds(-18000).unwrap()),
+    ///     time_zone_id: None,
+    ///     metazone_id: None,
+    ///     zone_variant: None,
+    /// });
+    ///
+    /// assert_eq!(tz_from_iana_annotation, CustomTimeZone {
+    ///     gmt_offset: None,
+    ///     time_zone_id: Some(TimeZoneBcp47Id(tinystr!(8, "uschi"))),
+    ///     metazone_id: None,
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:125]: Issue: I think you should fill in the `metazone_id`, which is computable from a `time_zone_id` and a local datetime, both of which which you have available in these IXDTF strings. See `MetazoneCalculator`.

Probably fine to leave it empty if you don't have a full datetime in the IXDTF string (e.g. if the time of day is missing).

File: components/timezone/src/ixdtf.rs [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+    ///
+    /// For more information on time zone parsing, see [`CustomTimeZone::try_from_ixdtf_str`].
+    #[cfg(feature = "compiled_data")]
+    pub fn try_from_ixdtf_str(ixdtf_str: &str) -> Result<Self, ParseError> {
+        Self::try_from_ixdtf_utf8(ixdtf_str.as_bytes())
+    }
+
+    /// Create a [`CustomZonedDateTime`] in any calendar from IXDTF syntax utf8 bytes.
+    ///
+    /// ✨ *Enabled with the `ixdtf` Cargo feature.*
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs:382]:
```suggestion
    /// ✨ *Enabled with the `compiled_data` and `ixdtf` Cargo features.*
```

**Review by sffc — COMMENTED** [review]
Nice work!

File: components/timezone/src/ixdtf.rs (hunk 1 above)
Comment (nekevss) [reply components/timezone/src/ixdtf.rs:69]: I have no specific case for keeping them. As a side note, there are probably some places in this PR that could be streamlined a bit more having the `try_from_ixdtf_record` publicly available in `icu_calendar`. At the same time though, I'm not entirely convinced it's worth it to expose those methods.

File: components/timezone/src/ixdtf.rs [hunk 11]
```diff
@@ -248,15 +262,34 @@ impl CustomTimeZone {
...
             TimeZoneRecord::Name(iana_identifier) => {
                 let mapper = TimeZoneIdMapper::new();
                 let bcp47_id = mapper
                     .as_borrowed()
                     .iana_bytes_to_bcp47(iana_identifier)
                     .ok_or(ParseError::InvalidIanaIdentifier)?;
-                Ok(Self::new_with_bcp47_id(bcp47_id))
+
+                let (year, month, day) =
+                    date.map(|d| (d.year, d.month, d.day)).unwrap_or((0, 0, 0));
```
Comment (sffc) [inline components/timezone/src/ixdtf.rs]: If the date or time is not present, then I think do not calculate the metazone. Only calculate it if all the fields are there.

**Review by sffc — APPROVED** [review]
